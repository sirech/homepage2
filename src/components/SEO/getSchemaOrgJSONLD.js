export default ({
  isBlogPost,
  url,
  siteUrl,
  title,
  author,
  image,
  siteTitle,
  description,
  datePublished,
}) => {
  const schemaOrgJSONLD = [
    {
      '@context': 'http://schema.org',
      '@type': 'WebSite',
      url,
      name: title,
      alternateName: siteTitle,
    },
  ]

  const post = {
    '@context': 'http://schema.org',
    '@type': 'BlogPosting',
    url,
    name: title,
    alternateName: siteTitle,
    headline: title,
    description,
    author: {
      '@type': 'Person',
      name: author,
    },
    publisher: {
      '@type': 'Organization',
      url: siteUrl,
      name: author,
    },
    mainEntityOfPage: {
      '@type': 'WebSite',
      '@id': siteUrl,
    },
    datePublished,
  }

  if (image) {
    post.image = image
  }

  return isBlogPost
    ? [
        ...schemaOrgJSONLD,
        {
          '@context': 'http://schema.org',
          '@type': 'BreadcrumbList',
          itemListElement: [
            {
              '@type': 'ListItem',
              position: 1,
              item: {
                '@id': url,
                name: title,
              },
            },
          ],
        },
        post,
      ]
    : schemaOrgJSONLD
}
