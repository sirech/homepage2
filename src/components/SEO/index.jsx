import React from 'react'
import { Helmet } from 'react-helmet'
import PropTypes from 'prop-types'

import { path as Rpath } from 'ramda'

import getSchemaOrgJSONLD from './getSchemaOrgJSONLD'

import siteType from '../../prop-types/site'

const buildDescription = (post) => {
  const description = Rpath(['frontmatter', 'description'])(post)
  return (
    description ||
    post.html
      .replace(/<figure>[\s\S]*<\/figure>/s, '')
      .replace(/<("[^"]*"|'[^']*'|[^'">])*>/g, '')
      .substr(0, 200)
  )
}

const SEO = ({ isBlogPost, post, site }) => {
  const title = Rpath(['frontmatter', 'title'])(post)
  const siteTitle = Rpath(['siteMetadata', 'title'])(site)
  const author = Rpath(['siteMetadata', 'author'])(site)

  const date = isBlogPost ? Rpath(['frontmatter', 'date'])(post) : false

  const description = buildDescription(post)

  const siteUrl = Rpath(['siteMetadata', 'url'])(site)
  const url = siteUrl + Rpath(['frontmatter', 'path'])(post)

  const draft = Rpath(['frontmatter', 'draft'])(post)
  const canonical = Rpath(['frontmatter', 'canonical'])(post)

  const image = Rpath([
    'frontmatter',
    'image',
    'childImageSharp',
    'fluid',
    'src',
  ])(post)
  const imageUrl = image ? siteUrl + image : null

  const schemaOrgJSONLD = getSchemaOrgJSONLD({
    isBlogPost,
    url,
    siteUrl,
    title,
    author,
    siteTitle,
    description,
    datePublished: date,
  })

  return (
    <Helmet title={`${title} | ${siteTitle}`}>
      {/* General tags */}
      <meta name="title" content={`${title} | ${siteTitle}`} />
      <meta name="description" content={description} />

      {/* Schema.org tags */}
      <script type="application/ld+json">
        {JSON.stringify(schemaOrgJSONLD)}
      </script>

      {/* hide drafts for google */}
      {draft ? (
        <meta name="robots" content="noindex" />
      ) : (
        <meta name="robots" content="index, follow" />
      )}

      {/* OpenGraph tags */}
      <meta property="og:url" content={url} />
      {isBlogPost ? <meta property="og:type" content="article" /> : null}
      <meta property="og:title" content={title} />
      <meta property="og:description" content={description} />
      {image && <meta property="og:image" content={imageUrl} />}

      {/* Twitter Card tags */}
      <meta
        name="twitter:card"
        content={image ? 'summary_large_image' : 'summary'}
      />
      {image && <meta name="twitter:image" content={imageUrl} />}
      <meta
        name="twitter:creator"
        content={Rpath(['siteMetadata', 'twitter'])(site)}
      />
      <meta name="twitter:title" content={title} />
      <meta name="twitter:description" content={description} />

      {/* Override canonical link */}
      {canonical ? (
        <link rel="canonical" key={canonical} href={canonical} />
      ) : null}
    </Helmet>
  )
}

SEO.propTypes = {
  isBlogPost: PropTypes.bool,
  post: PropTypes.shape({
    frontmatter: PropTypes.shape({
      categories: PropTypes.arrayOf(PropTypes.string),
      date: PropTypes.string.isRequired,
      layout: PropTypes.string,
      path: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      draft: PropTypes.bool.isRequired,
      canonical: PropTypes.string,
      image: PropTypes.shape({
        childImageSharp: PropTypes.shape({
          fluid: PropTypes.shape({
            src: PropTypes.string.isRequired,
          }),
        }),
      }),
    }).isRequired,
    html: PropTypes.string.isRequired,
  }).isRequired,
  site: siteType,
}

SEO.defaultProps = {
  isBlogPost: false,
}

export default SEO
