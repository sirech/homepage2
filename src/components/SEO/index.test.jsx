import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import SEO from './index'

describe('components', () => {
  describe('SEO', () => {
    const post = {
      html: '<p>this is the content</p>',
      frontmatter: {
        title: 'title',
        path: '/post',
        date: '2018/03/01',
        draft: false,
      },
    }

    const site = {
      meta: {
        author: 'me',
        description: 'description',
        title: 'site title',
        twitter: '@handle',
        url: 'http://example.com',
      },
    }

    it('renders correctly for a blogpost', () => {
      const component = shallow(<SEO isBlogPost post={post} site={site} />)
      expect(toJson(component)).toMatchSnapshot()
    })

    it('does not index draft posts', () => {
      const draft = {
        html: post.html,
        frontmatter: { ...post.frontmatter, draft: true },
      }

      const component = shallow(<SEO isBlogPost post={draft} site={site} />)
      expect(toJson(component)).toMatchSnapshot()
    })

    it('renders correctly for a normal page', () => {
      const component = shallow(
        <SEO isBlogPost={false} post={post} site={site} />
      )
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
