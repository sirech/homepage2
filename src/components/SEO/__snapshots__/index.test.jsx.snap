// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components SEO description can be overriden 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  title="My Post | site title"
>
  <meta
    content="My Post | site title"
    name="title"
  />
  <meta
    content="This is another description"
    name="description"
  />
  <script
    type="application/ld+json"
  >
    [{"@context":"http://schema.org","@type":"WebSite","url":"http://example.com/the-post","name":"My Post","alternateName":"site title"},{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"item":{"@id":"http://example.com/the-post","name":"My Post"}}]},{"@context":"http://schema.org","@type":"BlogPosting","url":"http://example.com/the-post","name":"My Post","alternateName":"site title","headline":"My Post","description":"This is another description","author":{"@type":"Person","name":"me"},"publisher":{"@type":"Organization","url":"http://example.com","name":"me"},"mainEntityOfPage":{"@type":"WebSite","@id":"http://example.com"}}]
  </script>
  <meta
    content="index, follow"
    name="robots"
  />
  <meta
    content="http://example.com/the-post"
    property="og:url"
  />
  <meta
    content="article"
    property="og:type"
  />
  <meta
    content="My Post"
    property="og:title"
  />
  <meta
    content="This is another description"
    property="og:description"
  />
  <meta
    content="summary"
    name="twitter:card"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="My Post"
    name="twitter:title"
  />
  <meta
    content="This is another description"
    name="twitter:description"
  />
</HelmetWrapper>
`;

exports[`components SEO does not index draft posts 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  title="My Post | site title"
>
  <meta
    content="My Post | site title"
    name="title"
  />
  <meta
    content="this is the content"
    name="description"
  />
  <script
    type="application/ld+json"
  >
    [{"@context":"http://schema.org","@type":"WebSite","url":"http://example.com/the-post","name":"My Post","alternateName":"site title"},{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"item":{"@id":"http://example.com/the-post","name":"My Post"}}]},{"@context":"http://schema.org","@type":"BlogPosting","url":"http://example.com/the-post","name":"My Post","alternateName":"site title","headline":"My Post","description":"this is the content","author":{"@type":"Person","name":"me"},"publisher":{"@type":"Organization","url":"http://example.com","name":"me"},"mainEntityOfPage":{"@type":"WebSite","@id":"http://example.com"}}]
  </script>
  <meta
    content="noindex"
    name="robots"
  />
  <meta
    content="http://example.com/the-post"
    property="og:url"
  />
  <meta
    content="article"
    property="og:type"
  />
  <meta
    content="My Post"
    property="og:title"
  />
  <meta
    content="this is the content"
    property="og:description"
  />
  <meta
    content="summary"
    name="twitter:card"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="My Post"
    name="twitter:title"
  />
  <meta
    content="this is the content"
    name="twitter:description"
  />
</HelmetWrapper>
`;

exports[`components SEO renders correctly for a blogpost 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  title="My Post | site title"
>
  <meta
    content="My Post | site title"
    name="title"
  />
  <meta
    content="this is the content"
    name="description"
  />
  <script
    type="application/ld+json"
  >
    [{"@context":"http://schema.org","@type":"WebSite","url":"http://example.com/the-post","name":"My Post","alternateName":"site title"},{"@context":"http://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"item":{"@id":"http://example.com/the-post","name":"My Post"}}]},{"@context":"http://schema.org","@type":"BlogPosting","url":"http://example.com/the-post","name":"My Post","alternateName":"site title","headline":"My Post","description":"this is the content","author":{"@type":"Person","name":"me"},"publisher":{"@type":"Organization","url":"http://example.com","name":"me"},"mainEntityOfPage":{"@type":"WebSite","@id":"http://example.com"}}]
  </script>
  <meta
    content="index, follow"
    name="robots"
  />
  <meta
    content="http://example.com/the-post"
    property="og:url"
  />
  <meta
    content="article"
    property="og:type"
  />
  <meta
    content="My Post"
    property="og:title"
  />
  <meta
    content="this is the content"
    property="og:description"
  />
  <meta
    content="summary"
    name="twitter:card"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="My Post"
    name="twitter:title"
  />
  <meta
    content="this is the content"
    name="twitter:description"
  />
</HelmetWrapper>
`;

exports[`components SEO renders correctly for a normal page 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  title="My Post | site title"
>
  <meta
    content="My Post | site title"
    name="title"
  />
  <meta
    content="this is the content"
    name="description"
  />
  <script
    type="application/ld+json"
  >
    [{"@context":"http://schema.org","@type":"WebSite","url":"http://example.com/the-post","name":"My Post","alternateName":"site title"}]
  </script>
  <meta
    content="index, follow"
    name="robots"
  />
  <meta
    content="http://example.com/the-post"
    property="og:url"
  />
  <meta
    content="My Post"
    property="og:title"
  />
  <meta
    content="this is the content"
    property="og:description"
  />
  <meta
    content="summary"
    name="twitter:card"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="My Post"
    name="twitter:title"
  />
  <meta
    content="this is the content"
    name="twitter:description"
  />
</HelmetWrapper>
`;
